lib/RAST_SDK/AnnotationUtils.pm

```diff
@@ -415,19 +415,65 @@ sub _check_NC_features {
     }
 }
 
-
 ## Mapping the contigIDs back to their original (long) names
 sub _remap_contigIDs {
-    my ($self, $contigID_hash, $genome) = @_;
+    my ($self, $contigID_hash, $gn) = @_;
+
+    return $gn unless $contigID_hash;
 
-    my $contigs = $genome->{contigs};
-    foreach my $ctg_obj (@{$contigs}) {
-        my $cid = $ctg_obj->{id};
-        $ctg_obj->{id} = $contigID_hash->{$cid};
-        $ctg_obj->{name} = $contigID_hash->{$cid};
+    if( $gn->{contig_ids} && @{$gn->{contig_ids}} > 0 ) {
+      for my $ctg_id (@{$gn->{contig_ids}}) {
+        if( $contigID_hash->{$ctg_id} ) {
+            $ctg_id = $contigID_hash->{$ctg_id};
+        }
+      }
     }
-    return $genome;
+    if( $gn->{contigs} && @{$gn->{contigs}} > 0) {
+      for my $ctg (@{$gn->{contigs}}) {
+        if( $contigID_hash->{$ctg->{id}} ) {
+            $ctg->{id} = $contigID_hash->{$ctg->{id}};
+		}
+        if( $contigID_hash->{$ctg->{name}} ) {
+            $ctg->{name} = $contigID_hash->{$ctg->{name}};
+        }
+      }
+    }
+    my $loc_ctg_id;
+    if($gn->{features} && @{$gn->{features}} > 0) {
+      for my $ftr (@{$gn->{features}}) {
+        $loc_ctg_id = $ftr->{location}[0][0];
+        if( $contigID_hash->{$loc_ctg_id} ) {
+            $loc_ctg_id = $contigID_hash->{$loc_ctg_id};
+        }
+      }
+    }
+    if( $gn->{non_coding_features} && @{$gn->{non_coding_features}} > 0) {
+      for my $nc_ftr (@{$gn->{non_coding_features}}) {
+        $loc_ctg_id = $nc_ftr->{location}[0][0];
+        if( $contigID_hash->{$loc_ctg_id} ) {
+            $loc_ctg_id = $contigID_hash->{$loc_ctg_id};
+        }
+      }
+    }
+    if( $gn->{cdss} && @{$gn->{cdss}} > 0) {
+      for my $cds (@{$gn->{cdss}}) {
+        $loc_ctg_id = $cds->{location}[0][0];
+        if( $contigID_hash->{$loc_ctg_id} ) {
+            $loc_ctg_id = $contigID_hash->{$loc_ctg_id};
+        }
+      }
+    }
+    if( $gn->{mrnas} && @{$gn->{mrnas}} > 0) {
+      for my $mrna (@{$gn->{mrnas}}) {
+        $loc_ctg_id = $mrna->{location}[0][0];
+        if( $contigID_hash->{$loc_ctg_id} ) {
+            $loc_ctg_id = $contigID_hash->{$loc_ctg_id};
+        }
+      }
+    }
```

Comment:
There's a lot of repetition here that could be refactored to make the method shorter and cleaner.

Reply:
Can you explain what this set of stanzas is doing? It isn't very clear from the code.

Reply:
I was trying to check the embeded contig ids one by one, from the remapping function results to the on-the-fly hash mapping.  Now instead of one by one, I put them into arrays and cmp_deeply of the arrays before and after remapping.